### PR TITLE
Integrated tests

### LINKED ISSUE #50: Regression tests
It would be really nice to create an "expected output" grid file and use it for regression testing. Maybe 3 tests - single null, connected double null and disconnected double null?

### PR DESCRIPTION
**Pull request by johnomotani**
Add integrated tests, based on an analytic, connected double-null equilibrium. These are regression tests, where output is compared to expected output. Getting the tests to pass on a different system that the one where the expected data was generated was challenging - some notes on making grid generation as reproducible as possible have been added on the wiki https://github.com/boutproject/hypnotoad/wiki/production.

There are tests for orthogonal and nonorthogonal grids, and also a second set similar to the first but where machine-precision level noise is added after the initial stage. The second set are intended to help in testing robustness of grid generation when generating new expected data. On the same system where the expected data is generated, the standard tests will pass even if e.g. tolerances are too loose. By adding some noise, we can check the robustness of the subsequent grid generation even on the system that generated the expected data.

To get the tests to pass on the Github CI server, also adds a number of small tweaks to the code to improve robustness of grid generation:
* When refining `PsiContour`s (after the first time) and when equalising spacing of `FineContour`s, the points at the start and end of the contour (not including 'extended' points past the target) are kept fixed. Since many iterations are sometimes necessary, previously the endpoints could drift slightly.
* Intersections with the wall are found from `FineContour`s rather than `PsiContour`s, which should improve accuracy (especially when `finecontour_Nfine` is made large).
* Start and end points of `FineContour`s are forced to be identical to the start and end points of the parent `PsiContour`, preventing different branches being followed due to small errors/differences in the positions.
* Some 'over-damping' (updating with a linear combination of the new value and old value at each iteration step) added to make iteration in `FineContour.equaliseSpacing` more robust.
* Several more tolerance options were added so that all tolerances can be made small when a reproducible grid is desired.
* Interpolation of positions from `FineContour` objects is now `"linear"` rather than `"cubic"` - this is more consistent with the way distances along the `FineContour`s are calculated.

Note: these changes will change the grids produced slightly when running from the same input file (although the changes should not be significant compared to the tolerances).

An extra, minor addition: the version numbers of Python and all loaded modules (excluding standard-library modules which are fixed by the Python version) are saved to the grid file.

- [x] Closes #50
- [x] Tests added
- [x] Updated `doc/whats-new.md` with a summary of the changes

### CODE REVIEW

File: hypnotoad/core/mesh.py [hunk 1]
```diff
@@ -2531,11 +2632,12 @@ def makeRegions(self):
                 ), "region should not have been added to any yGroup before"
                 region.yGroupIndex = len(group)
                 group.append(region)
-                region_set.remove(region)
+                region_list.pop(i)
```
Comment (ZedThree) [inline hypnotoad/core/mesh.py:2635]: I guess this is not a significant change from the previous code, but we should be very careful modifying lists while iterating over them. It might be better to make a copy and `pop` elements from that?

File: hypnotoad/core/mesh.py [hunk 2]
```diff
@@ -2531,11 +2632,12 @@ def makeRegions(self):
...
                 region = region.getNeighbour("upper")
                 if region is None or group.count(region) > 0:
                     # reached boundary or have all regions in a periodic group
                     break
+                i = region_list.index(region)
```
Comment (ZedThree) [inline hypnotoad/core/mesh.py:2640]: I can't see how `i` gets used after this, am I missing something?

**Review by ZedThree — APPROVED** [review]
LGTM, just a couple of minor questions/comments

Always good to see more tests! :tada:

File: hypnotoad/core/mesh.py [hunk 3]
```diff
@@ -701,22 +793,30 @@ def correct_sfunc_orthogonal(contour, sfunc_orthogonal_original):
...
+                # now make lower_intersect_index the index where the point at the wall
+                # is.
+                # check whether one of the points is close to the wall point and remove
+                # if it is - use a pretty loose checek because points apart from the
+                # wall point are about to be redistributed anyway, so does not matter if
+                # we remove one.
                 if (
                     calc_distance(contour[lower_intersect_index], lower_intersect)
-                    < self.atol
+                    < 1.0e-3
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply hypnotoad/core/mesh.py:804]: On second thoughts, I think a configurable option is the best thing. I'd thought this should never need changing, but conceivably there could be problems if points are too close together in the PsiContour (i.e. the threshold is too small) giving a bad interpolation function, or if they are too far apart (threshold too big) giving inaccurate interpolation.
